sophia/src/term/literal/_convert.rs

```diff
@@ -0,0 +1,495 @@
+//! Public exported by parent-module `literal`.
+//!
+
+use super::*;
+use crate::ns::xsd;
+use crate::term::{Iri, Result, StaticTerm, TermData, TermError};
+use crate::triple::stream::{SinkError, SourceError, StreamResult};
+
+/// Provide an IRI to specify the semantic datatype.
+pub trait DataType {
+    /// Static datatype of this Rust-type.
+    fn dt() -> &'static Iri<&'static str>;
+
+    /// Datatype of `self`.
+    ///
+    /// This can be used for `enum`s where each variant may has its own
+    /// datatype.
+    fn dt_of(&self) -> &'static Iri<&'static str> {
+        Self::dt()
+    }
+
+    fn dt_term() -> StaticTerm {
+        Iri::from(Self::dt()).into()
+    }
+
+    fn dt_term_of(&self) -> StaticTerm {
+        Iri::from(Self::dt_of(self)).into()
+    }
+}
+
+/// Representation as an RDF literal.
+pub trait AsLiteral<TD: TermData>: DataType {
+    /// Create an RDF literal, representing `self`.
+    fn as_literal(&self) -> Literal<TD>;
+
+    /// Blanked implementation to directly get a term.
+    fn as_term(&self) -> Term<TD> {
+        self.as_literal().into()
+    }
+}
+
+/// Create `self` from an RDF literal.
+pub trait FromLiteral: Sized {
+    /// Error raised if literal text is malformed.
+    type Error: std::error::Error;
+
+    /// Direct from a literal.
+    fn from_literal<TD: TermData>(lit: &Literal<TD>) -> Result<Self, Self::Error>;
+
+    /// Get a term that is actually a literal.
+    ///
+    /// # Error
+    ///
+    /// The blanked implementation throws `SourceError(UnexpectedKindOfTerm)`
+    /// if `t` is not a literal.
+    fn from_term<'a, TD>(t: &'a Term<TD>) -> StreamResult<Self, TermError, Self::Error>
+    where
+        TD: 'a + TermData,
+    {
+        if let Term::Literal(lit) = t {
+            Self::from_literal(lit).map_err(SinkError)
+        } else {
+            Err(SourceError(TermError::UnexpectedKindOfTerm {
+                term: t.to_string(),
+                expect: "literal".to_string(),
+            }))
+        }
+    }
+}
+
+macro_rules! impl_dt_with_term {
+    ($ty:ty, $term:expr) => {
+        impl $crate::term::literal::DataType for $ty {
+            fn dt() -> &'static $crate::term::iri::Iri<&'static str> {
+                if let Term::Iri(iri) = &$term {
+                    iri
+                } else {
+                    panic!("Should not have passed a non IRI")
+                }
+            }
+
+            fn dt_term() -> crate::term::Term<&'static str> {
+                $term.clone()
+            }
+        }
+    };
+}
+
+impl_dt_with_term!(u8, xsd::unsignedByte);
+impl_dt_with_term!(u16, xsd::unsignedShort);
+impl_dt_with_term!(u32, xsd::unsignedInt);
+impl_dt_with_term!(u64, xsd::unsignedLong);
+impl_dt_with_term!(i8, xsd::byte);
+impl_dt_with_term!(i16, xsd::short);
+impl_dt_with_term!(i32, xsd::integer);
```

Comment:
```suggestion
impl_dt_with_term!(i32, xsd::int);
```
`xsd:integer` is the common supertype of all types of integer. As you chose to use the most specific type for the others, it makes more sense to do the same here.